File: solutions/go/blackjack/1/blackjack.go
package blackjack

// ParseCard returns the integer value of a card following blackjack ruleset.
func ParseCard(card string) int {
	switch card {
        case "ace":		return 11
        case "two":		return 2
        case "three":	return 3
        case "four":	return 4
    	case "five":	return 5
    	case "six":		return 6
    	case "seven":	return 7
    	case "eight":	return 8
    	case "nine":	return 9
    	case "ten", "jack", "queen", "king": return 10
        default:		return 0
    }
}

// FirstTurn returns the decision for the first turn, given two cards of the
// player and one card of the dealer.
func FirstTurn(card1, card2, dealerCard string) string {
	playerScore := ParseCard(card1) + ParseCard(card2)
	dealerScore := ParseCard(dealerCard)

	switch {
    	// If you have a pair of aces, you must split (P)
    	case card1 == "ace" && card2 == "ace": return "P"
    
    	// Blackjack: Win (W) if dealer doesn't have an Ace or 10-value card
    	case playerScore == 21:
    		if dealerScore < 10 {
    			return "W"
    		}
    		return "S"
    
    	// Stand (S)
    	case playerScore >= 17 && playerScore <= 20: return "S"
    	case playerScore >= 12 && playerScore <= 16 && dealerScore < 7: return "S"
    
    	// Hit (H)
    	default: return "H"
	}
}

File: solutions/ruby/blackjack/1/blackjack.rb
module Blackjack
    @game_cards = {
      ace: 11,
      two: 2,
      three: 3,
      four: 4,
      five: 5,
      six: 6,
      seven: 7,
      eight: 8,
      nine: 9,
      ten: 10,
      jack: 10,
      queen: 10,
      king: 10
    }

  def self.parse_card(card)
    @game_cards[card.to_sym] || 0
  end

  def self.card_range(card1, card2)
    cards = [card1, card2]
    case cards.reduce(0) { |sum, card| sum + parse_card(card) }
      when 4..11  then "low"
      when 12..16 then "mid"
      when 7..20  then "high"
      when 21     then "blackjack"
    end
  end

  def self.first_turn(card1, card2, dealer_card)
    # Check for pair of aces - always split
    if card1 == "ace" && card2 == "ace"
      return "P"
    end
    
    player_sum = parse_card(card1) + parse_card(card2)
    dealer_value = parse_card(dealer_card)
    
    case player_sum
    when 21  # Blackjack
      # Win automatically unless dealer has ace, face card, or ten
      if ["ace", "ten", "jack", "queen", "king"].include?(dealer_card)
        "S"  # Stand and wait
      else
        "W"  # Automatic win
      end
    when 17..20  # Always stand
      "S"
    when 12..16  # Stand unless dealer has 7 or higher
      if dealer_value >= 7
        "H"  # Hit
      else
        "S"  # Stand
      end
    else  # 11 or lower - always hit
      "H"
    end
  end
end
